geo-traits/src/polygon.rs

```diff
@@ -0,0 +1,89 @@
+use crate::{Dimension, LineStringTrait, PolygonInteriorIterator};
+use geo_types::{CoordNum, LineString, Polygon};
+
+/// A trait for accessing data from a generic Polygon.
+pub trait PolygonTrait: Sized {
+    /// The coordinate type of this geometry
+    type T: CoordNum;
+
+    /// The type of each underlying ring, which implements [LineStringTrait]
+    type ItemType<'a>: 'a + LineStringTrait<T = Self::T>
+    where
+        Self: 'a;
+
+    /// The dimension of this geometry
+    fn dim(&self) -> Dimension;
+
+    /// The exterior ring of the polygon
+    fn exterior(&self) -> Option<Self::ItemType<'_>>;
+
+    /// An iterator of the interior rings of this Polygon
+    fn interiors(&self) -> PolygonInteriorIterator<'_, Self::T, Self::ItemType<'_>, Self> {
+        PolygonInteriorIterator::new(self, 0, self.num_interiors())
+    }
+
+    /// The number of interior rings in this Polygon
+    fn num_interiors(&self) -> usize;
+
+    /// Access to a specified interior ring in this Polygon
+    /// Will return None if the provided index is out of bounds
+    fn interior(&self, i: usize) -> Option<Self::ItemType<'_>> {
+        if i >= self.num_interiors() {
+            None
+        } else {
+            unsafe { Some(self.interior_unchecked(i)) }
+        }
+    }
+
+    /// Access to a specified interior ring in this Polygon
+    ///
+    /// # Safety
+    ///
+    /// Accessing an index out of bounds is UB.
+    unsafe fn interior_unchecked(&self, i: usize) -> Self::ItemType<'_>;
+}
+
+impl<T: CoordNum> PolygonTrait for Polygon<T> {
+    type T = T;
+    type ItemType<'a> = &'a LineString<Self::T> where Self: 'a;
+
+    fn dim(&self) -> Dimension {
+        Dimension::XY
+    }
+
+    fn exterior(&self) -> Option<Self::ItemType<'_>> {
+        // geo-types doesn't really have a way to describe an empty polygon
+        Some(Polygon::exterior(self))
+    }
+
+    fn num_interiors(&self) -> usize {
+        Polygon::interiors(self).len()
+    }
+
+    unsafe fn interior_unchecked(&self, i: usize) -> Self::ItemType<'_> {
+        unsafe { Polygon::interiors(self).get_unchecked(i) }
+    }
+}
+
+impl<'a, T: CoordNum> PolygonTrait for &'a Polygon<T> {
+    type T = T;
+    type ItemType<'b> = &'a LineString<Self::T> where
+        Self: 'b;
+
+    fn dim(&self) -> Dimension {
+        Dimension::XY
+    }
+
+    fn exterior(&self) -> Option<Self::ItemType<'_>> {
+        // geo-types doesn't really have a way to describe an empty polygon
+        Some(Polygon::exterior(self))
```

Comment:
null/empty geometries are a real "joy" of mine. (#1223, #1226)

I've always represented "POLYGON EMPTY" as a `geo::Polygon` whose exterior is a `LineString` with zero points. But maybe that thinking doesn't hold up across formats?

I expect some small code cleanups are possible when we don't have to check for None - but maybe there's downsides to that approach?

Asked differently: What things are made worse if you changed this signature to non-optional and returned an empty LineString rather than None?

Reply:
> I've always represented "POLYGON EMPTY" as a `geo::Polygon` whose exterior is a `LineString` with zero points. But maybe that thinking doesn't hold up across formats?

Oh, I did not expect that. I suppose we could go either way as long as it was well-documented, but to me, `None` is much clearer to say "empty" rather than zero-length LineString.

Reply:
Yeah the distinction feels mostly metaphysical. The decision predates me, so I think of it as mostly an accident of history at this point. Still though, it has its convenience — the null object pattern can be useful. 

Comparing LineString::empty() with to an Option< LineString>::None:

e.g. `None` doesn't have an length, or coordinate count (etc.), whereas an empty LineString has length (zero) and coordinate count (zero), so an empty LineString coalesces easily into most operations. Sometimes these kinds of things cause bugs though (#1226), though that bugs a little different - I don't think you're talking about representing `MultiPolygon(vec![])` as `Option<MultiPolygon>::None`, though maybe `MultiPolygon#polygons()` should return an Option::None rather than en empty list by your logic? Or is that different? 

The challenge with returning an Option is that it typically cascades through APIs which use it "infecting" anything that might interact with it. I guess there are those who think that's a feature - it should infect everything - forcing you to acknowledge it. It reminds me of wishing away NaN.

At the end of the day, it's still possible to build a polygon with an empty exterior (with geo anyway), so any code that wants to accept `geo::Polygon` has to handle it.

Reply:
> I don't think you're talking about representing `MultiPolygon(vec![])` as `Option<MultiPolygon>::None`, though maybe `MultiPolygon#polygons()` should return an Option::None rather than en empty list by your logic? Or is that different?

I see this as different because `PolygonTrait` has an explicit separation between `exterior()` and `interiors()`. If we have this separation, `exterior` says _here is this specific ring from the polygon_. In this case, I prefer `exterior()` to return `Option<Self::LineStringType>`. This is different from `MultiPolygonTrait` because it doesn't have a method to extract _a specific concept of its interior_, it yields _all its interior objects_.

So in summary, my argument is that if `PolygonTrait` had a `rings()` method that returned _both_ the exterior ring and any interior rings (with the documented assumption that the first ring is the exterior), then that should act the same as the `MultiPolygonTrait::polygon` method.

Essentially, the difference is that the other traits define a `num_*` method to refer to _all_ of its internal parts, while `PolygonTrait` has this distinction between exterior and interiors.

(Sorry if that was verbose, I struggled trying to articulate my thoughts)

Reply:
> So in summary, my argument is that if PolygonTrait had a rings() method that returned both the exterior ring and any interior rings (with the documented assumption that the first ring is the exterior), then that should act the same as the MultiPolygonTrait::polygon method.

I don't think it should have this method – we (not just GeoRust; I'm speaking generally) usually distinguish between `shell` and `holes`, `shell` being the outer linear ring, `holes` being the interior linear rings. I don't think a method that groups them together and expects the consumer to remember a convention is a good API design.

Reply:
I agree. I like the distinction between `exterior/shell` and `interiors/holes`. But in this case I prefer for `exterior()` to return an `Option<LineStringType>` rather than a `LineStringType` that consumers have to check has non-empty coordinates.